PP-261: Microsecond Logging

bhroam opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Bug/feature Description
* This PR implements the RFE for microsecond logging.  
*  [PP-261](https://pbspro.atlassian.net/browse/PP-261)
* PTL bug where the dshutils functions pbs_set_config(), pbs_unset_config(), and pbs_set_environment() would only work if the test was run as root.  Plus there wasn't pbs_unset_environment(), so that was implemented.
* Fixed test pbs_init_script.py to be able to be run as non-root.

#### Cause / Analysis / Design
* Design: https://pbspro.atlassian.net/wiki/spaces/PD/pages/50836942/PP-261+micro-second+time+stamp+for+daemon+logging
* Discussion: http://community.pbspro.org/t/pp-261-micro-second-time-stamp-for-daemon-logging/467

#### Solution Description
* PTL bug: The root cause was that all the functions called an internal function called _set_file().  _set_file() would call run_copy() without sudo=True.  I fixed it by adding a sudo argument to _set_file() which is passed to run_copy().  In all the affected functions, I pass sudo=True.
* Test bug: The test didn't call run_cmd() with sudo=True.  This meant the init script wouldn't be able to stop/start PBS.  Also, the env argument to run_cmd() won't work with the sudo argument.  It will pass the environmental variables to the sudo process which strips them from the real command.  I had to pass the environmental variables on the command line.
#### Testing logs/output
*  [ptl.log](https://github.com/PBSPro/pbspro/files/2033238/ptl.log)
* Other tests that use affected functions:  [ptl2.log](https://github.com/PBSPro/pbspro/files/2011259/ptl2.log)



#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [X] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [X] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [X] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [X] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [X] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [X] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [X] I have attached **test logs to this pull request** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

bhroam commented:
For some reason my pep8 and travis's pep8 report different things.  My pep8 reports cleanly, and travis fails.  This is causing a lot of commits.

bhroam commented:
Since no one had started my review yet, I just squashed all the pep8 fixing commits to make it easier to review.

agrawalravi90 commented on src/tools/tracejob.c:
```diff
@@ -481,13 +488,35 @@ parse_log(FILE *fp, char *job, int ind)
+					if (sscanf(tmp.date, "%d/%d/%d %d:%d:%d", &tms.tm_mon, &tms.tm_mday, 
+					    &tms.tm_year, &tms.tm_hour, &tms.tm_min, &tms.tm_sec) != 6) {
+						log_lines[ll_cur_amm].date_time = -1;	/* error in date field */
+						log_lines[ll_cur_amm].highres_ts = -1;
```
pull this line outside of if-else as it is common to both

agrawalravi90 commented on src/tools/tracejob.c:
```diff
@@ -481,13 +488,35 @@ parse_log(FILE *fp, char *job, int ind)
+ 						 * in the range 0 to 11 for mktime()
+ 						 */
+						tms.tm_mon--;
+						log_lines[ll_cur_amm].date_time = mktime(&tms);
```
Don;t you need to set 'log_lines[ll_cur_amm].highres_ts' here? I don't know what it's used for, but it's being set in other cases, so just checking

bhroam replied:
This is the case where highres_ts was set by the sscanf().  The other cases are either an error in the sscanf() or no microseconds were detected.  Those need to have it set to -1.

agrawalravi90 commented on src/tools/tracejob.c:
```diff
@@ -481,13 +488,35 @@ parse_log(FILE *fp, char *job, int ind)
+						log_lines[ll_cur_amm].date_time = -1;	/* error in date field */
+						log_lines[ll_cur_amm].highres_ts = -1;
+					} else {
+						has_high_res_timestamp = 1;
```
We have macros for True and False right? can you please use them instead of 1 and 0 literals?

bhroam replied:
It appears we don't have global macros for true and false.  There are some localized macros in different header files, but none of them are appropriate for tracejob to include.

agrawalravi90 commented on src/tools/tracejob.c:
```diff
@@ -481,13 +488,35 @@ parse_log(FILE *fp, char *job, int ind)
+					if (sscanf(tmp.date, "%d/%d/%d %d:%d:%d", &tms.tm_mon, &tms.tm_mday, 
+					    &tms.tm_year, &tms.tm_hour, &tms.tm_min, &tms.tm_sec) != 6) {
+						log_lines[ll_cur_amm].date_time = -1;	/* error in date field */
+						log_lines[ll_cur_amm].highres_ts = -1;
```
Does it make sense to make -1  a macro?

bhroam replied:
Probably.  It's set to -1 up front and then the -1 is checked as a value.  If we set it to -1 and then checked >= 0 that'd be one thing.  Since we are specifically looking for -1, we should use a macro.

agrawalravi90 commented on src/tools/tracejob.c:
```diff
@@ -563,6 +592,13 @@ sort_by_date(const void *v1, const void *v2)
 	else if (l1->date_time > l2->date_time)
 		return 1;
 	else {
+		if ((l1->highres_ts != -1) && (l2->highres_ts != -1)) {
```
same comment here, does it make sense to make -1 a macro?

agrawalravi90 commented on test/fw/ptl/utils/pbs_dshutils.py:
```diff
@@ -280,20 +285,33 @@ def _set_file(self, hostname, fin, fout, append, vars):
+            rv = self.run_copy(hostname, fn, fout, uid=user, gid=group,
+                               level=logging.DEBUG2, sudo=sudo)
+            if rv['rc'] != 0:
+                raise PbsConfigError
```
Would it be better to pass an error message to PbsConfigError saying that the copy failed?

bhroam replied:
It doesn't matter.  This is just really to trigger the try/except.  The except block does have an error message.  If I put one here, it would just be ignored.

bhroam commented on src/lib/Libifl/pbs_loadconf.c:
```diff
@@ -564,6 +565,10 @@ pbs_loadconf(int reload)
 			}
+			else if (!strcmp(conf_name, PBS_CONF_LOG_HIGHRES_TIMESTAMP)) {
+				if (sscanf(conf_value, "%u", &uvalue) == 1)
+					pbs_conf.pbs_log_highres_timestamp = ((uvalue > 0) ? 1 : 0);
```
It's for consistency.  pbs.conf booleans are 1 or 0.  This is how the code works for all of them.

arungrover commented on src/lib/Liblog/pbs_log.c:
```diff
@@ -858,7 +910,14 @@ log_record(int eventtype, int objclass, int sev, const char *objname, const char
+		microsec = tp.tv_usec;
+	} else {
+		now = time((time_t *)0);	/* get time for message */
+		microsec = 0;
```
You don't have to set it to 0 in this case. But, even if you keep it, I'm ok with it.

arungrover commented on src/tools/tracejob.h:
```diff
@@ -94,6 +94,7 @@ struct log_entry
 {
 	char *date;           /* date of log entry */
 	time_t date_time;     /* number of seconds from the epoch to date */
+	long highres_ts;  /* high resolution time stamp */
```
Please indent the comment. I think the variable name is misleading here. When we parse the log message microseconds are read into highres_ts which is part of a timestamp and not really a timestamp by itself.

arungrover commented on test/fw/ptl/utils/pbs_dshutils.py:
```diff
@@ -449,7 +472,46 @@ def set_pbs_environment(self, hostname=None,
+        :param hostname: the name of the host on which to operate
+        :type hostname: str or None
+        :param fin: the input pbs.conf file
+        :type fin: str or None
```
Default is not None, please change the comment

arungrover commented on test/tests/functional/pbs_highreslog.py:
```diff
@@ -0,0 +1,153 @@
+
+    def tearDown(self):
+        confs = self.du.parse_pbs_config()
+        if 'PBS_LOG_HIGHRES_TIMESTAMP' in confs:
```
Shouldn't this be added as part of default setup? why are you removing it here?

bhroam replied:
The problem is that pbs.conf is not reverted to default in the default setUp().  It needs to be removed here.

arungrover commented on test/tests/functional/pbs_highreslog.py:
```diff
@@ -0,0 +1,153 @@
+from tests.functional import *
+
+
+class HighResLoggingTestSuite(TestFunctional):
```
According to guideline test suites start with "Test" keyword

minghui-liu commented on test/fw/ptl/utils/pbs_dshutils.py:
```diff
@@ -356,18 +374,20 @@ def set_pbs_config(self, hostname=None, fin=None, fout=None,
         if fout is None and fin is not None:
             fout = fin
-        if confs:
+        if confs is not None:
```
`{}` is also not equal to `None`. Is this what you want?

bhroam replied:
It is what I want.  I changed the default value of confs from {} to None because dictionaries are mutable types.  Originally, if confs was defaulted and it was changed, that change would persist through all calls to set_pbs_config().  By defaulting it to None and then setting it to an empty dictionary in the code, you avoid this.

minghui-liu commented on test/tests/functional/pbs_highreslog.py:
```diff
@@ -0,0 +1,153 @@
+
+
+class HighResLoggingTestSuite(TestFunctional):
+
```
Please remove the empty line here.

bhroam commented:
I've addressed all comments except for the ones in the windows gettimeofday() function.  I didn't write it, so I'm a little scared to touch it.  I have emailed the people who did write it for assistance.  I'll address them in my next commit.

minghui-liu commented on test/tests/functional/pbs_highreslog.py:
```diff
@@ -98,17 +97,15 @@ def test_basic(self):
         self.du.set_pbs_config(confs=a, append=True)
         _msg = 'Failed to restart server: %s' % (self.server.shortname)
         self.assertTrue(self.server.restart(), _msg)
+        print self.du.parse_pbs_config()
```
Please use log instead of printing directly.

bhroam replied:
Or I can just delete it.  It's a debug message I forgot to remove :)

minghui-liu commented on src/lib/Liblog/pbs_log.c:
```diff
@@ -910,12 +910,13 @@ log_record(int eventtype, int objclass, int sev, const char *objname, const char
-		now = time((time_t *)0);	/* get time for message */
 
+	if (pbs_conf.pbs_log_highres_timestamp)
+		snprintf(microsec_buf, sizeof(microsec_buf), ".%0ld", (long)tp.tv_usec);
```
The indentation here looks off.

bhroam replied:
It's two tabs.  I think it is correct.

agrawalravi90 commented on src/lib/Liblog/pbs_log.c:
```diff
@@ -858,7 +901,13 @@ log_record(int eventtype, int objclass, int sev, const char *objname, const char
+	if (gettimeofday(&tp, NULL) != -1) {
+		now = tp.tv_sec;
+
+	if (pbs_conf.pbs_log_highres_timestamp)
```
The indentation seems off here

agrawalravi90 commented on src/lib/Liblog/pbs_log.c:
```diff
@@ -858,7 +901,13 @@ log_record(int eventtype, int objclass, int sev, const char *objname, const char
+	if (pbs_conf.pbs_log_highres_timestamp)
+		snprintf(microsec_buf, sizeof(microsec_buf), ".%06ld", (long)tp.tv_usec);
+	} else
+		now = time(NULL);
```
you do error checking for gettimeofday() but not for time(), why? I feel like we should just bail if gettimeofday fails.

agrawalravi90 commented:
Could you also fix Travis/Appveyor warnings? they've been failing for the last few commits.

bhroam commented:
Appvetor is failing due to an initialization issue.  It can't download one of its dependencies.  Hopefully it'll pass on my next push.

minghui-liu commented:
I restarted your travis and appveyor build.

bhroam commented:
OK I fixed the indentation and the pep8 issue.  As for gettimeofday() failing, instead of bailing and not printing the log message, I'll just print it at the epoch (Jan 1 1970 GMT).

agrawalravi90 commented on test/tests/functional/pbs_highreslog.py:
```diff
@@ -0,0 +1,148 @@
+        for line in lines:
+            m = self.tm_micro_re.match(line)
+            if m is not None:
+                ok = False
```
why not break from here?

bhroam replied:
I actually just turned this into a self.assertIsNone().  No need to break any more since the test will fail if the assert fails

agrawalravi90 commented on test/tests/functional/pbs_highreslog.py:
```diff
@@ -0,0 +1,148 @@
+               ' it shouldn\'t be there'
+        self.assertTrue(ok, _msg)
+        _msg = self.server.shortname + \
+            ': High resolution time stamp correctly not in log'
```
Did you mean "High resolution time stamp correctly not **set** in log"?

bhroam replied:
Thanks for pointing this out.  I fixed it.

agrawalravi90 commented on test/tests/functional/pbs_highreslog.py:
```diff
@@ -0,0 +1,148 @@
+        for line in lines:
+            m = self.tm_micro_re.match(line)
+            if m is not None:
+                ok = False
```
same here, shouldn't you break out?

bhroam replied:
Same as above, I replaced this with a self.assertIsNone()

agrawalravi90 commented on test/tests/functional/pbs_highreslog.py:
```diff
@@ -0,0 +1,148 @@
+            'it should not be there'
+        self.assertTrue(ok, _msg)
+        _msg = self.server.shortname + \
+            ': High resolution time stamp correctly not in tracejob output'
```
Again, did you mean to say "High resolution time stamp correctly not **set** in tracejob output" ?

bhroam replied:
Yep, same typo.  Thanks for pointing it out.

agrawalravi90 commented on test/tests/functional/pbs_init_script.py:
```diff
@@ -43,23 +43,28 @@ class TestPbsInitScript(TestFunctional):
+                'PBS_START_COMM': '1', 'PBS_START_MOM': '1'}
+        self.du.set_pbs_config(confs=conf)
+
+        self.du.run_cmd(cmd=['/etc/init.d/pbs', 'stop'], sudo=True)
```
why didn't you use PBSInitServices().stop() instead?

bhroam replied:
I don't want PBSInitServices() to muck with pbs.conf.  The whole point of this test suite is to test if environmental variables override pbs.conf variables.  I don't think PBSInitServices() will muck with pbs.conf unless you only stop one daemon, but I don't want to take the chance.

agrawalravi90 commented on test/tests/functional/pbs_init_script.py:
```diff
@@ -43,23 +43,28 @@ class TestPbsInitScript(TestFunctional):
-               'PBS_START_COMM': '0', 'PBS_START_MOM': '1'}
+        cmd = ['PBS_START_SERVER=0', 'PBS_START_SCHED=0',
+               'PBS_START_COMM=0', 'PBS_START_MOM=1',
+               '/etc/init.d/pbs', 'start']
```
Please use os.sep and os.path.join() to create the path

bhroam replied:
Good point.  I'll update this to use os.path.join()

agrawalravi90 commented on test/tests/functional/pbs_init_script.py:
```diff
@@ -43,23 +43,28 @@ class TestPbsInitScript(TestFunctional):
+
+        self.du.run_cmd(cmd=['/etc/init.d/pbs', 'stop'], sudo=True)
+
+        conf['START_PBS_MOM'] = '0'
```
Is START_PBS_MOM a typo? should it be PBS_START_MOM?

bhroam replied:
Good catch.  I tried to figure out why this test passed with this typo, and then I realized that PBS_START_MOM=1 from pbs.conf and worked fine by not being overridden.  I've corrected the typo.

agrawalravi90 commented on test/tests/selftest/pbs_dshutils_tests.py:
```diff
@@ -0,0 +1,85 @@
+from ptl.utils.pbs_dshutils import PbsConfigError
+
+
+class TestDshUtils(TestSelf):
```
Thanks for writing these unit tests! I hope this inspires others to do the same.

agrawalravi90 commented:
Thanks for posting the latest logs, you have my approval.

agrawalravi90 commented:
@arungrover could you please sign-off on the latest changes as well so that this can be merged?

agrawalravi90 commented:
@bhroam please rebase for the merge
